tokio-boring/src/async_callbacks.rs

```diff
@@ -0,0 +1,148 @@
+use boring::ex_data::Index;
+use boring::ssl::{self, ClientHello, Ssl, SslContextBuilder};
+use once_cell::sync::Lazy;
+use std::future::Future;
+use std::pin::Pin;
+use std::task::{ready, Context, Poll};
+
+type BoxSelectCertFuture = Pin<
+    Box<
+        dyn Future<Output = Result<BoxSelectCertFinish, AsyncSelectCertError>>
+            + Send
+            + Sync
+            + 'static,
+    >,
+>;
+
+type BoxSelectCertFinish =
+    Box<dyn FnOnce(ClientHello<'_>) -> Result<(), AsyncSelectCertError> + 'static>;
+
+pub(crate) static TASK_CONTEXT_INDEX: Lazy<Index<Ssl, usize>> =
```

Comment:
I don't like how I use `usize` here but it's not that important, as other parts of boring already do to store a `SslContext` pointer in `Ssl`. We should probably use a pointer wrapper that implements `Send` and `Sync` for convenience at some point.